File: project_euler_unified/problem_697/solution.py
#!/usr/bin/env python3
"""Project Euler Problem 697: Randomly Decaying Sequence"""

from scipy.stats import norm, chi2
import math

def solve():
    n = 10_000_000
    nu = 2 * n  # degrees of freedom for chi-squared

    # We need P(X_n < 1) = 0.25
    # X_n = c * prod(U_i), -ln(U_i) ~ Exp(1)
    # -2*sum(ln(U_i)) ~ chi^2(2n)
    # P(X_n < 1) = P(sum(-ln(U_i)) > ln(c)) = P(chi^2(2n) > 2*ln(c)) = 0.25
    # So 2*ln(c) = chi^2 quantile at 0.75

    chi2_val = chi2.ppf(0.75, nu)
    log10_c = chi2_val / (2 * math.log(10))

    print(f"{log10_c:.2f}")

    z_075 = norm.ppf(0.75)  # ~0.6744897501960817
    term = 1 - 2/(9*nu) + z_075 * math.sqrt(2/(9*nu))
    chi2_approx = nu * term**3
    log10_c_approx = chi2_approx / (2 * math.log(10))
    print(f"Wilson-Hilferty: {log10_c_approx:.2f}")

solve()

File: project_euler_unified/problem_697/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    // We need P(X_n < 1) = 0.25 where X_n = c * prod(U_i)
    // This gives 2*ln(c) = chi^2_{2n, 0.75}
    // Using Wilson-Hilferty approximation for chi-squared quantile

    long long n = 10000000LL;
    double nu = 2.0 * n;

    // z_{0.75} = inverse normal CDF at 0.75
    // Using the known value
    double z = 0.6744897501960817;

    // Wilson-Hilferty approximation
    double term = 1.0 - 2.0 / (9.0 * nu) + z * sqrt(2.0 / (9.0 * nu));
    double chi2 = nu * term * term * term;

    // 2*ln(c) = chi2, so ln(c) = chi2/2
    // log10(c) = ln(c) / ln(10) = chi2 / (2 * ln(10))
    double log10c = chi2 / (2.0 * log(10.0));

    // Round to 2 decimal places
    cout << fixed << setprecision(2) << log10c << endl;

    return 0;
}
